homeassistant/components/sensor/dsmr.py

```diff
@@ -0,0 +1,153 @@
+"""
+Support for Dutch Smart Meter Requirements.
+
+Also known as: Smartmeter or P1 port.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.dsmr/
+"""
+
+import asyncio
+import logging
+
+import voluptuous as vol
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_DEVICE
+from homeassistant.helpers.entity import Entity
+
+DOMAIN = 'dsmr'
+
+REQUIREMENTS = ['dsmr-parser==0.2']
+
+CONF_DSMR_VERSION = 'dsmr_version'
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_DSMR_VERSION = '4'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+    vol.Optional(CONF_DSMR_VERSION, default=DEFAULT_DSMR_VERSION): vol.All(
+        cv.string, vol.In(['4', '2.2'])),
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup DSMR sensors."""
+    from dsmr_parser.obis_references import (
+        CURRENT_ELECTRICITY_USAGE,
+        CURRENT_ELECTRICITY_DELIVERY,
+        ELECTRICITY_ACTIVE_TARIFF
+    )
+
+    devices = []
+
+    dsmr = DSMR(hass, config, devices)
+
+    devices += [
+        DSMREntity('Power Usage', CURRENT_ELECTRICITY_USAGE, dsmr),
+        DSMREntity('Power Production', CURRENT_ELECTRICITY_DELIVERY, dsmr),
+        DSMRTariff('Power Tariff', ELECTRICITY_ACTIVE_TARIFF, dsmr),
+    ]
+    yield from async_add_devices(devices, True)
+    yield from dsmr.async_update()
+
+
+class DSMR:
+    """DSMR interface."""
+
+    def __init__(self, hass, config, devices):
+        """Setup DSMR serial interface and add device entities."""
+        from dsmr_parser.serial import (
+            SERIAL_SETTINGS_V4,
+            SERIAL_SETTINGS_V2_2,
+            SerialReader
+        )
+        from dsmr_parser import telegram_specifications
+
+        dsmr_versions = {
+            '4': (SERIAL_SETTINGS_V4, telegram_specifications.V4),
+            '2.2': (SERIAL_SETTINGS_V2_2, telegram_specifications.V2_2),
+        }
+
+        device = config[CONF_DEVICE]
+        dsmr_version = config[CONF_DSMR_VERSION]
+
+        self.dsmr_parser = SerialReader(
+            device=device,
+            serial_settings=dsmr_versions[dsmr_version][0],
+            telegram_specification=dsmr_versions[dsmr_version][1],
+        )
+
+        self.hass = hass
+        self.devices = devices
+        self._telegram = {}
+
+    @asyncio.coroutine
+    def async_update(self):
```

Comment:
@aequitas if you're planning on making `dsmr-parser` then your current implementation will be appropriate.